src/java.base/share/classes/java/util/zip/GZIPHeaderBuilder.java

```diff
@@ -0,0 +1,361 @@
+                crc.update(filenameBytes, 0, filenameBytes.length);
+                crc.update(0);
+            }
+        }
+        // write file comment.
+        if ((flags & GZIPHeaderData.FCOMMENT) == GZIPHeaderData.FCOMMENT) {
+            /*
+             *    +===================================+
+             *    |...file comment, zero-terminated...|
+             *    +===================================+
+             */
+            byte[] fcommBytes = fileComment.getBytes("ISO-8859-1");
+            baos.write(fcommBytes);
+            baos.write(0);
+            if ((flags & GZIPHeaderData.FHCRC) == GZIPHeaderData.FHCRC) {
+                crc.update(fcommBytes, 0, fcommBytes.length);
+                crc.update(0);
+            }
+        }
+        // write header crc16.
+        if ((flags & GZIPHeaderData.FHCRC) == GZIPHeaderData.FHCRC) {
+            int crc16 = (int) crc.getValue() & 0xffff;
+            this.headerCrc16 = crc16;
+            baos.write((byte) (crc16 & 0xff));
+            baos.write((byte) (crc16 >> 8) & 0xff);
+            crc.reset();
+        }
+        return baos.toByteArray();
+    }
+
+    /**
+     * Verify extra field data layout.
+     * Per RFC 1952:
+     * If the FEXTRA bit is set, an "extra field" is present in
+     * the header, with total length XLEN bytes.  It consists of a
+     * series of subfields, each of the form:
+     *
+     * +---+---+---+---+==================================+
+     * |SI1|SI2|  LEN  |... LEN bytes of subfield data ...|
+     * +---+---+---+---+==================================+
+     *
+     * SI1 and SI2 provide a subfield ID, typically two ASCII letters
+     * with some mnemonic value
+     *
+     * @param fieldBytes    the data of extra fileds.
+     * @return {@code true} if field data layout is correct.
+     *
+     * @since 18
+     */
+    private boolean verifyExtraFieldLayout(byte[] fieldBytes) {
+        int index = 0;
+        int total = fieldBytes.length;
+        while(index < total) {
+            byte si1 = fieldBytes[index++];
+            byte si2 = fieldBytes[index++];
+            // si1 and si2 should be ASCII
+            if (si1 > 127 || si2 > 127) return false;
+            byte loLen = fieldBytes[index++];
+            byte hiLen = fieldBytes[index++];
+            int len = ((int)hiLen) << 8 | (int)loLen;
+            index += len;
+        }
+        return index == total;
+    }
+
+    /**
+     * This class implements the header of GZIP file which contains members defined
+     * in the RFC 1952 specification
+     *
+     * @since 18
+     *
+     */
+    public record GZIPHeaderData (byte compressMethod, byte flags,
+                                  byte[] extraFieldBytes,
+                                  String filename,
+                                  String fileComment,
+                                  int headerCRC16,
+                                  byte[] headerBytes) {
```

Comment:
Need to override the getter for byte array fields to return copies than original arrays to prevent modifications. For the extraFieldBytes one, the call to copy needs a null check too.